core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.js

```diff
@@ -206,41 +223,59 @@ oppia.factory('UnitsObjectFactory', [function() {
     return unitsWithMultiplier;
   };
 
-  var unitToDict = function(unitsWithMultiplier) {
-    var unitDict = {};
+  var unitToList = function(unitsWithMultiplier) {
+    var unitList = [];
+    var flag = 0;
     for (var i = 0; i < unitsWithMultiplier.length; i++) {
       var unit = unitsWithMultiplier[i][0];
       var multiplier = unitsWithMultiplier[i][1];
       var ind = unit.indexOf('^');
+      flag = 0;
       if (ind > -1) {
         var s = unit.substr(0, ind);
         var power = parseInt(unit.substr(ind + 1));
       } else {
         var s = unit;
         var power = 1;
       }
-      if (!(s in unitDict)) {
-        unitDict[s] = 0;
+      for (var j = 0; j < unitList.length; j++) {
```

Comment:
This is inefficient. I recommend creating a dict just like earlier code and then convert that dict to a unitList by placing the key of dict in unit and value in exp. This way you can avoid entire O(n) wasted in this nested loop.

Reply:
Done!